taskchampion/taskchampion/src/server/cloud/server.rs

```diff
@@ -0,0 +1,1122 @@
+use super::service::{Service, ObjectInfo};
+#[cfg(not(test))]
+use std::time::{SystemTime, UNIX_EPOCH};
+use crate::errors::{Error, Result};
+use crate::server::crypto::{Cryptor, Sealed, Unsealed};
+use crate::server::{
+    AddVersionResult, GetVersionResult, HistorySegment, Server, Snapshot, SnapshotUrgency,
+    VersionId,
+};
+use ring::rand;
+use std::collections::{HashMap, HashSet};
+use uuid::Uuid;
+
+/// Implement the Server trait for a cloud service implemented by [`Service`].
+///
+/// This type implements a TaskChampion server over a basic object-storage service. It encapsulates
+/// all of the logic to ensure a linear sequence of versions, encrypt and decrypt data, and clean
+/// up old data so that this can be supported on a variety of cloud services.
+///
+/// ## Encryption
+///
+/// The encryption scheme is described in `sync-protocol.md`. The salt value used for key
+/// derivation is the SHA256 hash of the string "TaskChampion". Object names are not encrypted,
+/// by the nature of key/value stores. Since the content of the "latest" object can usually be
+/// inferred from object names, it, too, is not encrypted.
```

Comment:
I wonder if this will be a point of contention!

Encrypting this is a little difficult, since we're using AEAD, with a version_id as the additional data. That additional data is not encrypted, but I don't see a way to extract it in the `ring:aead` API. For snapshots and versions, the `version_id` is clear, but for "latest" it is not. We could pick a fixed value (say, `Uuid::max()`), but I'm not sure that's wise, cryptographically.

I'm also not sure that encrypting the value of "latest" accomplishes anything:
 - As noted in the docstring, in most cases "latest" can be determined by just looking at the version object names -- the version without a child is probably latest. So encrypting this value is not hiding anything from someone with access to the bucket.
 - Thinking in terms of integrity, does encrypting "latest" help? Assume Mal has read/write access to the bucket. An unencrypted "latest" could be replaced with a value pointing to an old version, or a nonexistent version, both of which would break sync for all replicas using the bucket. But Mal could also just copy the value of "latest", wait a few days, and then restore that old value, to accomplish the same thing. Even if that was made impossible, Mal could just delete version objects and accomplish the same thing. In short, if Mal has write access to the bucket then it's trivial to break sync, regardless of whether "latest" is encrypted.

Reply:
BTW I learned that I misread the description of AEAD. The additional data is not encrypted and does not appear anywhere in the output, so there's no way to "extract" it.